lib/zlib.js

```diff
@@ -466,164 +449,134 @@ function emitCloseNT(self) {
+  constructor(opts) {
+    super(opts, constants.DEFLATE);
   }
+}
 
-  this._processChunk(chunk, flushFlag, cb);
-};
-
-Zlib.prototype._processChunk = function(chunk, flushFlag, cb) {
-  var availInBefore = chunk && chunk.length;
-  var availOutBefore = this._chunkSize - this._offset;
-  var inOff = 0;
-
-  var self = this;
-
-  var async = typeof cb === 'function';
-
-  if (!async) {
-    var buffers = [];
-    var nread = 0;
+class Inflate extends Zlib {
+  constructor(opts) {
+    super(opts, constants.INFLATE);
+  }
+}
 
-    var error;
-    this.on('error', function(er) {
-      error = er;
-    });
+class Gzip extends Zlib {
+  constructor(opts) {
+    super(opts, constants.GZIP);
+  }
+}
 
-    assert(this._handle, 'zlib binding closed');
-    do {
-      var res = this._handle.writeSync(flushFlag,
-                                       chunk, // in
-                                       inOff, // in_off
-                                       availInBefore, // in_len
-                                       this._buffer, // out
-                                       this._offset, //out_off
-                                       availOutBefore); // out_len
-    } while (!this._hadError && callback(res[0], res[1]));
-
-    if (this._hadError) {
-      throw error;
-    }
+class Gunzip extends Zlib {
+  constructor(opts) {
+    super(opts, constants.GUNZIP);
+  }
+}
 
-    if (nread >= kMaxLength) {
-      _close(this);
-      throw new RangeError(kRangeErrorMessage);
-    }
+class DeflateRaw extends Zlib {
+  constructor(opts) {
+    super(opts, constants.DEFLATERAW);
+  }
+}
 
-    var buf = Buffer.concat(buffers, nread);
-    _close(this);
+class InflateRaw extends Zlib {
+  constructor(opts) {
+    super(opts, constants.INFLATERAW);
+  }
+}
 
-    return buf;
+class Unzip extends Zlib {
+  constructor(opts) {
+    super(opts, constants.UNZIP);
   }
+}
 
-  assert(this._handle, 'zlib binding closed');
-  var req = this._handle.write(flushFlag,
-                               chunk, // in
-                               inOff, // in_off
-                               availInBefore, // in_len
-                               this._buffer, // out
-                               this._offset, //out_off
-                               availOutBefore); // out_len
-
-  req.buffer = chunk;
-  req.callback = callback;
-
-  function callback(availInAfter, availOutAfter) {
-    // When the callback is used in an async write, the callback's
-    // context is the `req` object that was created. The req object
-    // is === this._handle, and that's why it's important to null
-    // out the values after they are done being used. `this._handle`
-    // can stay in memory longer than the callback and buffer are needed.
-    if (this) {
-      this.buffer = null;
-      this.callback = null;
-    }
+function createFn(type) {
+  const fn = function(o) {
+    return new type(o);
+  };
+  Object.defineProperty(fn, Symbol.hasInstance, {
```

Comment:
Instead of defining `Symbol.hasInstance`, it might be better to set the actual `fn.prototype` to `type.prototype`